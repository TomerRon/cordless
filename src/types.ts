import { ApplicationCommandOptionAllowedChannelTypes } from '@discordjs/builders'
import { APIApplicationCommandOptionChoice } from 'discord-api-types/v10'
import Discord, {
  ApplicationCommandOptionType,
  ClientEvents,
  ClientOptions,
  CommandInteraction,
} from 'discord.js'

/** Initialization options for your cordless bot */
export type InitOptions<C extends CustomContext = {}> = {
  /** The commands used by your bot. */
  commands?: BotCommand<C>[]
  /** The functions used by your bot */
  functions: BotFunction<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
  /**
   * Your bot token.
   *
   * @see https://discordjs.guide/preparations/setting-up-a-bot-application.html#your-bot-s-token
   */
  token: string
  /** A custom context object which will extend the context passed to your commands and functions */
  context?: C
  /**
   * Override the default Gateway Intents of the discord.js client.
   *
   * By default, the discord.js client will initialize with the [GUILDS, GUILD_MESSAGES] intents.
   * These default intents should be sufficient in most cases.
   *
   * @see https://discord.com/developers/docs/topics/gateway#gateway-intents
   * @see https://discordjs.guide/popular-topics/intents.html
   */
  intents?: ClientOptions['intents']
  /**
   * Generate a help function for the bot, which will be triggered by the value.
   *
   * For example, given a value of "!help", the following commands will be available:
   *
   * - !help (shows a list of available functions)
   * - !help <function-name> (shows a description and usage instructions for the given function)
   */
  helpCommand?: string
}

export type BotCommand<C extends CustomContext = {}> =
  | BotCommandWithHandler<C>
  | BotCommandWithSubcommands<C>

export interface BotCommandWithHandler<C extends CustomContext = {}>
  extends BotCommandBase {
  handler: (args: BotCommandHandlerArgs<C>) => void | Promise<void>
  options?: BotCommandOption[]
  subcommands?: never
}

export interface BotCommandWithSubcommands<C extends CustomContext = {}>
  extends BotCommandBase {
  subcommands: BotCommandWithHandler<C>[]
  handler?: never
  options?: never
}

type BotCommandBase = {
  name: string
  description?: string
}

export type BotCommandHandlerArgs<C extends CustomContext = {}> = {
  interaction: CommandInteraction
  context: Context<C>
}

export type BotCommandOption =
  | BotCommandStringOption
  | BotCommandIntegerOption
  | BotCommandBooleanOption
  | BotCommandUserOption
  | BotCommandChannelOption
  | BotCommandRoleOption
  | BotCommandMentionableOption
  | BotCommandNumberOption
  | BotCommandAttachmentOption

export interface BotCommandStringOption extends BotCommandOptionBase {
  type: 'STRING'
  choices?: APIApplicationCommandOptionChoice<string>[]
}

export interface BotCommandIntegerOption extends BotCommandOptionBase {
  type: 'INTEGER'
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandBooleanOption extends BotCommandOptionBase {
  type: 'BOOLEAN'
}

export interface BotCommandUserOption extends BotCommandOptionBase {
  type: 'USER'
}

export interface BotCommandChannelOption extends BotCommandOptionBase {
  type: 'CHANNEL'
  channelTypes?: ApplicationCommandOptionAllowedChannelTypes[]
}

export interface BotCommandRoleOption extends BotCommandOptionBase {
  type: 'ROLE'
}

export interface BotCommandMentionableOption extends BotCommandOptionBase {
  type: 'MENTIONABLE'
}

export interface BotCommandNumberOption extends BotCommandOptionBase {
  type: 'NUMBER'
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandAttachmentOption extends BotCommandOptionBase {
  type: 'ATTACHMENT'
}

type BotCommandOptionBase = {
  type: ApplicationCommandOptionType
  name: string
  description?: string
  required?: boolean
}

export type BotFunction<
  E extends keyof ClientEvents = 'messageCreate',
  C extends CustomContext = {},
> = {
  /**
   * The name of this function (no spaces allowed).
   * It will be displayed in the help function if there is a helpCommand.
   */
  name?: string
  /**
   * The description of this function and usage instructions.
   * It will be displayed in the help function if there is a helpCommand.
   */
  description?: string
  /**
   * The event this function should subscribe to (default: "messageCreate").
   */
  event?: E
  /** Determines whether or not this function should run */
  condition: (...args: [...ClientEvents[E], Context<C>]) => boolean
  /** Called whenever this function should run */
  callback: (
    ...args: [...ClientEvents[E], Context<C>]
  ) => void | Promise<Discord.Message | void>
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomContext = Record<string, any>

export type Context<C extends CustomContext = {}> = {
  client: Discord.Client<true>
  functions: BotFunction<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
} & C
