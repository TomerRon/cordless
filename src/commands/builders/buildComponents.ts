import { CommandInteraction, MessageActionRow, MessageButton } from 'discord.js'
import { BotCommandWithHandler, Context, CustomContext } from '../../types'

interface BuildComponentsArgs<C extends CustomContext = {}> {
  command: BotCommandWithHandler<C>
  interaction: CommandInteraction
  context: Context<C>
}

/**
 * Builds a list of message components for a given BotCommandWithHandler.
 */
const buildComponents = async <C extends CustomContext = {}>({
  command,
  interaction,
  context,
}: BuildComponentsArgs<C>): Promise<MessageActionRow[] | undefined> => {
  const { components } = command

  if (!components?.length) return

  const buttons: MessageButton[] = []

  for (const [i, component] of components.entries()) {
    const { label, style = 'PRIMARY' } = component

    if (component.style === 'LINK') {
      const url =
        typeof component.url === 'string'
          ? component.url
          : await component.url({ interaction, context })

      buttons.push(
        new MessageButton().setLabel(label).setStyle(style).setURL(url),
      )

      continue
    }

    const customId = `${command.name}-${label}-${i}`

    buttons.push(
      new MessageButton().setCustomId(customId).setLabel(label).setStyle(style),
    )
  }

  const row = new MessageActionRow().addComponents(buttons)

  return [row]
}

export default buildComponents
