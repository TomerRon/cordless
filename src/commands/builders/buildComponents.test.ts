import { Client, CommandInteraction } from 'discord.js'
import { BotCommandComponent, BotCommandWithHandler } from '../../types'
import buildComponents from './buildComponents'

const mockMessageActionRow = {
  addComponents: jest.fn().mockReturnThis(),
}

const mockMessageButton = {
  setCustomId: jest.fn().mockReturnThis(),
  setLabel: jest.fn().mockReturnThis(),
  setStyle: jest.fn().mockReturnThis(),
  setURL: jest.fn().mockReturnThis(),
}

jest.mock('discord.js', () => {
  const originalModule = jest.requireActual('discord.js')

  return {
    ...originalModule,
    MessageActionRow: jest.fn().mockImplementation(() => mockMessageActionRow),
    MessageButton: jest.fn().mockImplementation(() => mockMessageButton),
  }
})

describe('buildComponents', () => {
  beforeEach(jest.clearAllMocks)

  const mockCommand: BotCommandWithHandler = {
    name: 'command-a',
    description: 'command-a-desc',
    handler: jest.fn(),
  }

  const mockInteraction = {} as unknown as CommandInteraction

  const mockContext = {
    client: jest.fn() as unknown as Client,
    functions: [],
    foo: 'bar',
  }

  describe('when the command has no components', () => {
    it('does nothing', async () => {
      expect(
        await buildComponents({
          command: mockCommand,
          interaction: mockInteraction,
          context: mockContext,
        }),
      ).toBeUndefined()
    })
  })

  describe('when the command has an empty list of components', () => {
    it('does nothing', async () => {
      expect(
        await buildComponents({
          command: { ...mockCommand, components: [] },
          interaction: mockInteraction,
          context: mockContext,
        }),
      ).toBeUndefined()
    })
  })

  describe('when the command has one interactive button', () => {
    const mockComponent: BotCommandComponent = {
      label: 'component-a-label',
      style: 'PRIMARY',
      handler: jest.fn(),
    }

    it('builds a row with the component', async () => {
      expect(
        await buildComponents({
          command: { ...mockCommand, components: [mockComponent] },
          interaction: mockInteraction,
          context: mockContext,
        }),
      ).toStrictEqual([mockMessageActionRow])

      expect(mockMessageButton.setCustomId).toHaveBeenCalledWith(
        `command-a-component-a-label-0`,
      )
      expect(mockMessageButton.setLabel).toHaveBeenCalledWith(
        mockComponent.label,
      )
      expect(mockMessageButton.setStyle).toHaveBeenCalledWith(
        mockComponent.style,
      )
      expect(mockMessageButton.setURL).not.toHaveBeenCalled()

      expect(mockMessageActionRow.addComponents).toHaveBeenCalledWith([
        mockMessageButton,
      ])
    })
  })

  describe('when the command has one link button with a string url', () => {
    const mockComponent: BotCommandComponent = {
      label: 'component-a-label',
      style: 'LINK',
      url: 'component-url',
    }

    it('builds a row with the component', async () => {
      expect(
        await buildComponents({
          command: { ...mockCommand, components: [mockComponent] },
          interaction: mockInteraction,
          context: mockContext,
        }),
      ).toStrictEqual([mockMessageActionRow])

      expect(mockMessageButton.setCustomId).not.toHaveBeenCalled()
      expect(mockMessageButton.setLabel).toHaveBeenCalledWith(
        mockComponent.label,
      )
      expect(mockMessageButton.setStyle).toHaveBeenCalledWith(
        mockComponent.style,
      )
      expect(mockMessageButton.setURL).toHaveBeenCalledWith(mockComponent.url)

      expect(mockMessageActionRow.addComponents).toHaveBeenCalledWith([
        mockMessageButton,
      ])
    })
  })

  describe('when the command has one link button with a url resolve callback', () => {
    const mockResolvedUrl = 'component-url'
    const mockComponent: BotCommandComponent = {
      label: 'component-a-label',
      style: 'LINK',
      url: jest.fn().mockResolvedValue(mockResolvedUrl),
    }

    it('builds a row with the component', async () => {
      expect(
        await buildComponents({
          command: { ...mockCommand, components: [mockComponent] },
          interaction: mockInteraction,
          context: mockContext,
        }),
      ).toStrictEqual([mockMessageActionRow])

      expect(mockMessageButton.setCustomId).not.toHaveBeenCalled()
      expect(mockMessageButton.setLabel).toHaveBeenCalledWith(
        mockComponent.label,
      )
      expect(mockMessageButton.setStyle).toHaveBeenCalledWith(
        mockComponent.style,
      )
      expect(mockMessageButton.setURL).toHaveBeenCalledWith(mockResolvedUrl)

      expect(mockMessageActionRow.addComponents).toHaveBeenCalledWith([
        mockMessageButton,
      ])

      expect(mockComponent.url).toHaveBeenCalledWith({
        interaction: mockInteraction,
        context: mockContext,
      })
    })
  })

  describe('snapshots', () => {
    const mockComponentA: BotCommandComponent = {
      label: 'component-a-label',
      style: 'PRIMARY',
      handler: jest.fn(),
    }

    const mockComponentB: BotCommandComponent = {
      label: 'component-b-label',
      style: 'LINK',
      url: 'component-c-url',
    }

    const mockComponentC: BotCommandComponent = {
      label: 'component-c-label',
      handler: jest.fn(),
    }

    const mockComponentD: BotCommandComponent = {
      label: 'component-d-label',
      style: 'LINK',
      url: () => Promise.resolve('component-c-url'),
    }

    const mockComponentE: BotCommandComponent = {
      label: 'component-e-label',
      style: 'DANGER',
      handler: jest.fn(),
    }

    const mockComponents = [
      mockComponentA,
      mockComponentB,
      mockComponentC,
      mockComponentD,
      mockComponentE,
    ]

    it('matches the snapshot', async () => {
      expect(
        await buildComponents({
          command: { ...mockCommand, components: mockComponents },
          interaction: mockInteraction,
          context: mockContext,
        }),
      ).toStrictEqual([mockMessageActionRow])

      expect(mockMessageActionRow.addComponents).toHaveBeenCalledWith([
        mockMessageButton,
        mockMessageButton,
        mockMessageButton,
        mockMessageButton,
        mockMessageButton,
      ])

      Object.entries(mockMessageButton).forEach(([key, fn]) =>
        expect(fn.mock.calls).toMatchSnapshot(key),
      )
    })
  })
})
