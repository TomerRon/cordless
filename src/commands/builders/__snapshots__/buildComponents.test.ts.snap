// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildComponents snapshots matches the snapshot: setCustomId 1`] = `
Array [
  Array [
    "command-a-component-a-label-0",
  ],
  Array [
    "command-a-component-c-label-2",
  ],
  Array [
    "command-a-component-e-label-4",
  ],
]
`;

exports[`buildComponents snapshots matches the snapshot: setLabel 1`] = `
Array [
  Array [
    "component-a-label",
  ],
  Array [
    "component-b-label",
  ],
  Array [
    "component-c-label",
  ],
  Array [
    "component-d-label",
  ],
  Array [
    "component-e-label",
  ],
]
`;

exports[`buildComponents snapshots matches the snapshot: setStyle 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    5,
  ],
  Array [
    1,
  ],
  Array [
    5,
  ],
  Array [
    4,
  ],
]
`;

exports[`buildComponents snapshots matches the snapshot: setURL 1`] = `
Array [
  Array [
    "component-c-url",
  ],
  Array [
    "component-c-url",
  ],
]
`;
